transport/src/main/java/io/netty/channel/pool/FixedChannelPool.java

```diff
@@ -0,0 +1,142 @@
+import io.netty.util.concurrent.Promise;
+
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static io.netty.util.internal.ObjectUtil.checkNotNull;
+
+/**
+ * {@link ChannelPool} implementation that takes another {@link ChannelPool} implementation and enfore a maximum
+ * number of concurrent connections.
+ *
+ * @param <C>   the {@link Channel} type to pool.
+ * @param <K>   the {@link ChannelPoolKey} that is used to store and lookup the {@link Channel}s.
+ */
+public final class FixedChannelPool<C extends Channel, K extends ChannelPoolKey> implements ChannelPool<C, K> {
+
+    private final AtomicInteger acquiredChannelCount = new AtomicInteger();
+    private final Queue<AcquireTask<C, K>> pendingAcquireQueue = new ConcurrentLinkedQueue<AcquireTask<C, K>>();
+    private final FutureListener<C> decrementListener = new FutureListener<C>() {
+        @Override
+        public void operationComplete(Future<C> future) throws Exception {
+            if (future.isSuccess()) {
+                future.getNow().closeFuture().addListener(closeListener);
+            } else {
+                acquiredChannelCount.decrementAndGet();
+                runTaskQueue();
+            }
+        }
+    };
+    private final FutureListener<Boolean> runListener = new FutureListener<Boolean>() {
+        @Override
+        public void operationComplete(Future<Boolean> future) throws Exception {
+            runTaskQueue();
+        }
+    };
+    private final ChannelFutureListener closeListener = new ChannelFutureListener() {
+        @Override
+        public void operationComplete(ChannelFuture future) throws Exception {
+            // The channel was closed so try to process the taskqueue again as this means we lost one connection out
+            // of the pool.
+            runTaskQueue();
+        }
+    };
+
+    private final ChannelPool<C, K> pool;
+    private final int maxConnections;
+
+    public FixedChannelPool(ChannelPool<C, K> pool, int maxConnections) {
+        if (maxConnections < 1) {
+            throw new IllegalArgumentException("maxConnections must be >= 1 but was " + maxConnections);
+        }
+        this.pool = checkNotNull(pool, "pool");
+        this.maxConnections = maxConnections;
+    }
+
+    @Override
+    public Future<C> acquire(K key) {
+        return acquire(key, ImmediateEventExecutor.INSTANCE.<C>newPromise());
+    }
+
+    @Override
+    public Future<Boolean> release(C channel) {
+        return release(channel, ImmediateEventExecutor.INSTANCE.<Boolean>newPromise());
+    }
+
+    @Override
+    public Future<C> acquire(K key, Promise<C> promise) {
+        if (acquiredChannelCount.incrementAndGet() <= maxConnections) {
+            promise.addListener(decrementListener);
+            return pool.acquire(key, promise);
+        }
+        pendingAcquireQueue.add(new AcquireTask<C, K>(key, promise));
```

Comment:
@pjulien because the queue is MPSC and not MPSP. That said after we I would move the runTaskQueue() to get executed by the same thread all the time we could use it.